codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java

```diff
@@ -111,38 +122,42 @@ public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf
             int payloadLength = data.readableBytes() + padding + flags.getPaddingPresenceFieldLength();
             verifyPayloadLength(payloadLength);
 
-            ByteBuf out = ctx.alloc().buffer(FRAME_HEADER_LENGTH + payloadLength);
-
-            writeFrameHeader(out, payloadLength, DATA, flags, streamId);
-
-            writePaddingLength(padding, out);
+            buf = ctx.alloc().buffer(DATA_FRAME_HEADER_LENGTH);
+            writeFrameHeaderInternal(buf, payloadLength, DATA, flags, streamId);
+            writePaddingLength(buf, padding);
+            ctx.write(buf, promiseAggregator.newPromise());
+        } catch (Throwable t) {
```

Comment:
Can we combine the two try/catch blocks and just have a boolean tracking whether we need to release the data buffer?